arbor/util/spatial_tree.hpp

```diff
@@ -0,0 +1,181 @@
+template <typename T, std::size_t DIM>
+class spatial_tree {
+public:
+    static_assert(DIM >= 1, "Dimension of tree must be at least 1.");
+
+    using value_type = T;
+    using point_type = std::array<double, DIM>;
+    using node_data = std::vector<spatial_tree>;
+    using leaf_data = std::vector<T>;
+    using location_func_type = point_type (*)(const T &);
+
+    spatial_tree(): size_(0), data_(leaf_data()) {}
+
+    // Create a tree of given maximum depth and target leaf size. If any leaf holds more than the
+    // target size, it is recursively split into up to 2^DIM nodes until reaching the maximum depth.
+    // The "location" function type must have signature (const T&) -> point_type.
+    spatial_tree(std::size_t max_depth,
+        std::size_t leaf_size_target,
+        leaf_data data,
+        location_func_type location):
+        size_(data.size()),
+        data_(std::move(data)),
+        location_(location) {
+        auto &leaf_d = std::get<leaf_data>(data_);
+        if (leaf_d.empty()) return;
+
+        min_.fill(std::numeric_limits<double>::max());
+        max_.fill(-std::numeric_limits<double>::max());
+
+        for (const auto &d: leaf_d) {
+            const auto p = location(d);
+            for (std::size_t i = 0; i < DIM; ++i) {
+                if (p[i] < min_[i]) min_[i] = p[i];
+                if (p[i] > max_[i]) max_[i] = p[i];
+            }
+        }
+
+        point_type mid;
+        for (std::size_t i = 0; i < DIM; ++i) { mid[i] = (max_[i] - min_[i]) / 2.0 + min_[i]; }
+
+        if (max_depth > 1 && leaf_d.size() > leaf_size_target) {
+            constexpr auto divisor = math::pow<std::size_t, std::size_t>(2, DIM);
+
+            // The initial index of the sub node containing p
+            auto sub_node_index = [&](const point_type &p) {
+                std::size_t index = 0;
+                for (std::size_t i = 0; i < DIM; ++i) { index += i * 2 * (p[i] >= mid[i]); }
+                return index;
+            };
+
+            node_data new_nodes;
+            new_nodes.reserve(divisor);
+
+            // assign each point to sub-node
+            std::array<leaf_data, divisor> new_leaf_data;
+            for (const auto &d: leaf_d) {
+                const auto p = location(d);
+                new_leaf_data[sub_node_index(p)].emplace_back(d);
+            }
+
+            // move data into new sub-nodes if not empty
+            for (auto &l_d: new_leaf_data) {
+                if (l_d.size())
+                    new_nodes.emplace_back(max_depth - 1, leaf_size_target, std::move(l_d), location);
+            }
+
+            // replace current data_ with new sub-nodes
+            this->data_ = std::move(new_nodes);
+        }
+    }
+
+    spatial_tree(const spatial_tree &) = default;
+
+    spatial_tree(spatial_tree &&t) { *this = std::move(t); }
+
+    spatial_tree &operator=(const spatial_tree &) = default;
+
+    spatial_tree &operator=(spatial_tree &&t) {
```

Comment:
I wonder whether `std::swap`ping with `t` might be better. At least it avoids issues with exceptions.

Reply:
Using `std::swap` now for the container type now.

Reply:
After another look at it, I think using std::swap doesn't really at a benefit here, since we end up using move assignment anyway. It should be noexcept however.